# Анализ результатов выборов в г. Краснодар

<h3>Описание данных</h3>


***В распоряжении три датасета:***

- **Таблица report_uik (состав всех УИК):**


    - фио — ФИО члена комиссии;
    - статус — статус члена комиссии;
    - партия — партия члена комиссии;
    - уик — номер УИК;
    - адрес — адрес УИК.
    
- **Таблицы  report_result_part_1 и report_result_part_2 (результаты выборов в разрезе УИК):**


    - различные показатели результатов выборов в разрезе УИК.
    

*Данные сайта www.krasnodar.vybory.izbirkom.ru/region/krasnodar*

<h3> Этапы исследования </h3>

1. Импорт и предобработка данных.
2. Анализ данных:

    - Задача 1
    - Задача 2


3. Выводы.

In [1]:
import pandas as pd

In [2]:
import warnings
warnings.filterwarnings('ignore')

<h3>1. Импорт и предобработка данных</h3>

In [3]:
# Импорт составов всех УИК (файл report_uik.csv - сформирован с помощью парсинга, см. файл - парсинг_данных.ipynb)

report_uik=pd.read_csv('report_uik.csv')

In [4]:
# Импорт результатов выборов в разрезе УИК (файлы report_result_part_1.csv и report_result_part_2.csv сформированы
# с помощью выгрузки с сайта www.krasnodar.vybory.izbirkom.ru/region/krasnodar сводных таблиц)

result_1=pd.read_csv('report_result_part_1.csv',sep=';', encoding='cp1251')
result_2=pd.read_csv('report_result_part_2.csv',sep=';', encoding='cp1251')

In [5]:
result_1=result_1.rename(columns={'Unnamed: 0':'№_УИК'})
result_1=result_1.dropna(subset=['№_УИК'])
result_1 = result_1.set_index(['№_УИК'])
result_tr_1=result_1.transpose()

In [6]:
result_2=result_2.rename(columns={'Unnamed: 0':'№_УИК'})
result_2=result_2.dropna(subset=['№_УИК'])
result_2 = result_2.set_index(['№_УИК'])
result_tr_2=result_2.transpose()

In [7]:
result_tr=pd.concat([result_tr_1,result_tr_2])

In [8]:
result=result_tr.drop(['Район'], axis=1)
result=result.astype('int64')
result['Район']=result_tr['Район']

In [9]:
result.info()

<class 'pandas.core.frame.DataFrame'>
Index: 354 entries, 2001 to 2356
Data columns (total 19 columns):
 #   Column                                                                   Non-Null Count  Dtype 
---  ------                                                                   --------------  ----- 
 0   Число избирателей, внесенных в список                                    354 non-null    int64 
 1   Число бюллетеней, полученных участковой комиссией                        354 non-null    int64 
 2   Число бюллетеней, выданных избирателям, проголосовавшим досрочно         354 non-null    int64 
 3   Число бюллетеней, выданных участковой комиссией избирателям в помещении  354 non-null    int64 
 4   Число бюллетеней, выданных избирателям, проголосовавшим вне помещения    354 non-null    int64 
 5   Число погашенных бюллетеней                                              354 non-null    int64 
 6   Число бюллетеней, содержащихся в переносных ящиках                       354 non-nu

In [10]:
result.duplicated().sum()

0

In [11]:
result.head()

№_УИК  Число избирателей, внесенных в список  \
2001                                    2517   
2002                                    2164   
2003                                     969   
2004                                    2592   
2005                                    2316   

№_УИК  Число бюллетеней, полученных участковой комиссией  \
2001                                                2100   
2002                                                1800   
2003                                                 900   
2004                                                2100   
2005                                                2100   

№_УИК  Число бюллетеней, выданных избирателям, проголосовавшим досрочно  \
2001                                                 469                  
2002                                                 813                  
2003                                                 582                  
2004                                                 451                  
2005                                                 683                  

№_УИК  Число бюллетеней, выданных участковой комиссией избирателям в помещении  \
2001                                                 292                         
2002                                                 394                         
2003                                                  17                         
2004                                                 389                         
2005                                                 239                         

№_УИК  Число бюллетеней, выданных избирателям, проголосовавшим вне помещения  \
2001                                                  23                       
2002                                                   0                       
2003                                                   0                       
2004                                                  30                       
2005                                                   8                       

№_УИК  Число погашенных бюллетеней  \
2001                          1316   
2002                           593   
2003                           301   
2004                          1230   
2005                          1170   

№_УИК  Число бюллетеней, содержащихся в переносных ящиках  \
2001                                                 243    
2002                                                 813    
2003                                                   0    
2004                                                 219    
2005                                                 685    

№_УИК  Число бюллетеней, содержащихся в стационарных ящиках  \
2001                                                 541      
2002                                                 389      
2003                                                 599      
2004                                                 631      
2005                                                 245      

№_УИК  Число недействительных бюллетеней  Число действительных бюллетеней  \
2001                                  24                              760   
2002                                  18                             1184   
2003                                  17                              582   
2004                                  19                              831   
2005                                  11                              919   

№_УИК  Число утраченных бюллетеней  \
2001                             0   
2002                             0   
2003                             0   
2004                             0   
2005                             0   

№_УИК  Число бюллетеней, не учтенных при получении  \
2001                                             0   
2002                                             0   
2003                                             0   
2004                                      

**Таблица result - итоговая таблица с результатами выбровов в разрезе УИК. Пропусков и дубликатов в данных нет.**

In [12]:
report_uik=report_uik.drop(['Unnamed: 0', 'Unnamed: 1'], axis=1)
report_uik.columns=['фио', 'статус',
       'партия', 'уик', 'адрес']

In [13]:
report_uik.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5261 entries, 0 to 5260
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   фио     5261 non-null   object
 1   статус  5261 non-null   object
 2   партия  5261 non-null   object
 3   уик     5261 non-null   object
 4   адрес   5261 non-null   object
dtypes: object(5)
memory usage: 205.6+ KB


In [14]:
report_uik.head()

фио            статус  \
0           Ермакова Нина Сергеевна      Председатель   
1        Рязанцева Елена Викторовна  Зам.председателя   
2     Чебуракова Наталья Викторовна         Секретарь   
3          Андрианов Михаил Юрьевич              Член   
4  Богомолов Александр Владимирович              Член   

                                              партия  \
0               собрание избирателей по месту работы   
1               собрание избирателей по месту работы   
2               собрание избирателей по месту работы   
3  Краснодарское местное отделение Краснодарского...   
4               собрание избирателей по месту работы   

                                       уик  \
0  Участковая избирательная комиссия №2301   
1  Участковая избирательная комиссия №2301   
2  Участковая избирательная комиссия №2301   
3  Участковая избирательная комиссия №2301   
4  Участковая избирательная комиссия №2301   

                                               адрес  
0  350011, Краснодарский край, город Краснодар, Ц...  
1  350011, Краснодарский край, город Краснодар, Ц...  
2  350011, Краснодарский край, город Краснодар, Ц...  
3  350011, Краснодарский край, город Краснодар, Ц...  
4  350011, Краснодарский край, город Краснодар, Ц...

In [15]:
report_uik.duplicated().sum()

0

**Таблица report_uik - итоговая таблица с составами УИК. Пропусков и дубликатов в данных нет.**

<h3>2. Анализ</h3>

**Задача 1.**

In [16]:
report_uik['партия_КПРФ']=report_uik['партия'].str.contains('КОММУНИСТИЧЕСКАЯ', na=False)

In [17]:
report_uik.loc[report_uik['партия_КПРФ']==True,'группы_партий']='КПРФ'
report_uik['группы_партий']=report_uik['группы_партий'].fillna('Другие')

In [18]:
print('Число представителей КПРФ и других партий:')
print()
report_uik['группы_партий'].value_counts()

Число представителей КПРФ и других партий:



Другие    4955
КПРФ       306
Name: группы_партий, dtype: int64

In [19]:
uik_n=report_uik['уик'].str.split('\№',expand=True)
uik_n.columns=['УК','№_уик']
report_uik = pd.concat([report_uik,uik_n[['№_уик']]],axis=1)

In [20]:
pivot_uik=report_uik.pivot_table(index='№_уик', columns='группы_партий', values='фио', aggfunc='count').reset_index()
pivot_uik=pivot_uik.merge(result[['Округ номер (ОИК)','Район']], how='left', left_on='№_уик', right_index=True)

df_without_kprf=pivot_uik[pivot_uik['КПРФ'].isna()][['№_уик','Округ номер (ОИК)','Район']]
list_without_kprf=df_without_kprf['№_уик'].tolist()

In [21]:
print('Количество УИК без представителей КПРФ: {} из {} участков'
      .format(df_without_kprf['№_уик'].count(),pivot_uik['№_уик'].count()))
print()
print('Номера УИК, в составе которых не было представителя КПРФ:')
print()
display(df_without_kprf)
display(df_without_kprf.pivot_table(index='Район', values='№_уик', aggfunc='count'))
print('Номера УИК в виде списка: ')
print()
print(list_without_kprf)

Количество УИК без представителей КПРФ: 35 из 341 участков

Номера УИК, в составе которых не было представителя КПРФ:



№_уик  Округ номер (ОИК)         Район
3    2004                1.0      Западный
6    2007                1.0      Западный
17   2018                2.0      Западный
20   2021                3.0      Западный
25   2026                4.0      Западный
33   2034                3.0      Западный
61   2101                8.0   Карасунский
86   2126               15.0   Карасунский
134  2174               17.0   Карасунский
140  2180               16.0   Карасунский
186  2239               22.0  Прикубанский
210  2263               24.0  Прикубанский
211  2264               18.0  Прикубанский
213  2266               22.0  Прикубанский
282  6011               27.0   Калининский
283  6012               27.0   Калининский
284  6013               27.0   Калининский
292  6021               32.0   Калининский
302  6031               30.0   Калининский
305  6034               31.0   Калининский
307  6036               29.0   Калининский
308  6037               29.0   Калининский
320  6049               28.0   Калининский
323  6052               30.0   Калининский
324  6053               31.0   Калининский
325  6054               29.0   Калининский
326  6055               28.0   Калининский
327  6056               27.0   Калининский
328  6057               27.0   Калининский
329  6058               27.0   Калининский
331  6060               30.0   Калининский
334  6063               33.0   Калининский
335  6064               33.0   Калининский
336  6065               29.0   Калининский
337  6066               29.0   Калининский

№_уик
Район              
Западный          6
Калининский      21
Карасунский       4
Прикубанский      4

Номера УИК в виде списка: 

['2004', '2007', '2018', '2021', '2026', '2034', '2101', '2126', '2174', '2180', '2239', '2263', '2264', '2266', '6011', '6012', '6013', '6021', '6031', '6034', '6036', '6037', '6049', '6052', '6053', '6054', '6055', '6056', '6057', '6058', '6060', '6063', '6064', '6065', '6066']


**Вывод по задаче 1:** В 35 УИК из 341 участка нет представителей КПРФ. Больше всего УИК без представителя КПРФ в Калининском районе (21 участок), в центральном районе в каждом УИК есть представитель КПРФ.   

**Задача 2.**

In [22]:
result_kprf=result[['Число действительных бюллетеней', 'КПРФ']]
result_kprf.columns=['бюллетеней_всего', 'за_КПРФ']
result_kprf.index= result_kprf.index.astype('int64')

In [23]:
result_kprf['группы_участков']=result_kprf.index.where(result_kprf.index.isin(list_without_kprf), 'участки_c_КПРФ')
result_kprf['группы_участков']=result_kprf['группы_участков'].where(result_kprf['группы_участков']=='участки_c_КПРФ', 'участки_без_КПРФ')
#result_kprf.head()

In [24]:
pivot_kprf=result_kprf.pivot_table(index='группы_участков', 
                        values=['бюллетеней_всего', 'за_КПРФ'], aggfunc='sum')

In [25]:
pivot_kprf['%_голосов_за_КПРФ']=((pivot_kprf['за_КПРФ']/pivot_kprf['бюллетеней_всего'].sum())*100).round(2)

In [26]:
# Приводим к запрошенному формату вывода

pivot_kprf_tr=pivot_kprf.transpose()
pivot_kprf_tr['delta']=pivot_kprf_tr['участки_c_КПРФ']-pivot_kprf_tr['участки_без_КПРФ']
pivot_kprf_tr

группы_участков    участки_c_КПРФ  участки_без_КПРФ     delta
бюллетеней_всего        191230.00          23433.00  167797.0
за_КПРФ                  42263.00           4064.00   38199.0
%_голосов_за_КПРФ           19.69              1.89      17.8

*Для того чтобы можно было видеть размер групп ("участки с КПРФ" и "без КПРФ"), в выводе решил оставить абсолютные значения голосов. Группы могут быть не сопостовимы по размеру и тогда сравнение относительных величин может быть некорректным.*

**В данном случае группы по размеру сопоставимы и можно сделать вывод, что в группе, где в составе комиссии был представитель КПРФ процент голосов за КПРФ (19.69%) в десять раза больше, чем в группе, где представителя этой партии не было (1.89%).** 

<h3>3. Выводы</h3>

**Вывод по задаче 1:** 

В 35 УИК из 341 участка нет представителей КПРФ. Больше всего УИК без представителя КПРФ в Калининском районе (21 участок), в центральном районе в каждом УИК есть представитель КПРФ.

Список УИК без представителей КПРФ: ['2004', '2007', '2018', '2021', '2026', '2034', '2101', '2126', '2174', '2180', '2239', '2263', '2264', '2266', '6011', '6012', '6013', '6021', '6031', '6034', '6036', '6037', '6049', '6052', '6053', '6054', '6055', '6056', '6057', '6058', '6060', '6063', '6064', '6065', '6066']

**Вывод по задаче 2:** 

В группе, где в составе комиссии был представитель КПРФ процент голосов за КПРФ (19.69%) в десять раза больше, чем в группе, где представителя этой партии не было (1.89%). Дельта - 17.8%